pkg/connection/reconcile_geo_replication.go

```diff
@@ -0,0 +1,205 @@
+// limitations under the License.
+
+package connection
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/streamnative/pulsar-resources-operator/api/v1alpha1"
+	resourcev1alpha1 "github.com/streamnative/pulsar-resources-operator/api/v1alpha1"
+	"github.com/streamnative/pulsar-resources-operator/pkg/admin"
+	"github.com/streamnative/pulsar-resources-operator/pkg/reconciler"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// PulsarGeoReplicationReconciler reconciles a PulsarGeoReplication object
+type PulsarGeoReplicationReconciler struct {
+	conn *PulsarConnectionReconciler
+	log  logr.Logger
+}
+
+func makeGeoReplicationReconciler(r *PulsarConnectionReconciler) reconciler.Interface {
+	return &PulsarGeoReplicationReconciler{
+		conn: r,
+		log:  r.log.WithName("PulsarGeoReplication"),
+	}
+}
+
+// Observe checks the updates of object
+func (r *PulsarGeoReplicationReconciler) Observe(ctx context.Context) error {
+	geoList := &resourcev1alpha1.PulsarGeoReplicationList{}
+	if err := r.conn.client.List(ctx, geoList, client.InNamespace(r.conn.connection.Namespace),
+		client.MatchingFields(map[string]string{
+			".spec.connectionRef.name": r.conn.connection.Name,
+		})); err != nil {
+		return fmt.Errorf("list GeoReplication [%w]", err)
+	}
+	r.log.V(1).Info("Observed geo replication items", "Count", len(geoList.Items))
+
+	r.conn.geoReplications = geoList.Items
+	if !r.conn.hasUnreadyResource {
+		for i := range r.conn.geoReplications {
+			if !resourcev1alpha1.IsPulsarResourceReady(&r.conn.geoReplications[i]) {
+				r.conn.hasUnreadyResource = true
+				break
+			}
+		}
+	}
+
+	r.log.V(1).Info("Observe Done")
+	return nil
+}
+
+// Observe checks the updates of object
+func (r *PulsarGeoReplicationReconciler) Reconcile(ctx context.Context) error {
+	for i := range r.conn.geoReplications {
+		geoReplication := &r.conn.geoReplications[i]
+		if err := r.ReconcileGeoReplication(ctx, r.conn.pulsarAdmin, geoReplication); err != nil {
+			return fmt.Errorf("reconcile geo replication [%w]", err)
+		}
+	}
+	return nil
+}
+
+func (r *PulsarGeoReplicationReconciler) ReconcileGeoReplication(ctx context.Context, pulsarAdmin admin.PulsarAdmin,
+	geoReplication *resourcev1alpha1.PulsarGeoReplication) error {
+
+	log := r.log.WithValues("pulsargeoreplication", geoReplication.Name, "namespace", geoReplication.Namespace)
+	log.V(1).Info("Start Reconcile")
+
+	destConnection := &resourcev1alpha1.PulsarConnection{}
+	namespacedName := types.NamespacedName{
+		Name:      geoReplication.Spec.DestinationConnectionRef.Name,
+		Namespace: geoReplication.Namespace,
+	}
+	if err := r.conn.client.Get(ctx, namespacedName, destConnection); err != nil {
+		log.Error(err, "Failed to get destination connection for geo replication")
+		if apierrors.IsNotFound(err) {
```

Comment:
Should this be `!apierrors.IsNotFound(err)` and return nil if it is?

Reply:
Fixed.